Initial Implementation of Kotlin SDK

wba2hi opened the pull request:
Initial Implementation of the Java / Kotlin SDK for Velocitas. 

Covering: 
- Base classes for service discovery (ClientFactories, via environment variables)
- VehicleApp base class (lifecycle management)
- Generated code: Selected grpc library; gRPC connection management
- Logger

Misc:
- Kotest Integration + Kotest BehaviorSpec Tests
- .editorconfig integration
- .gitignore 

Not covered: 

AsyncGrpcFacade: 
- https://github.com/eclipse-velocitas/vehicle-app-cpp-sdk/blob/main/sdk/include/sdk/grpc/AsyncGrpcFacade.h
- https://github.com/eclipse-velocitas/vehicle-app-cpp-sdk/blob/main/sdk/src/sdk/grpc/AsyncGrpcFacade.cpp
=> ClientContext is a cpp-specific grpc class used for "configuration", nothing comparable exists for the Java specific grpc implementation 

Out-of-Scope: 
- Vehicle Model (VSS classes) → future versions
- Communication to vehicle data broker via grpc → future versions
- MQTT-related topics → future versions
- Reliable messaging library and boilerplate code → future versions
- No Android-app-template → separate story

wba2hi committed:
Initial Implementation of Kotlin SDK

wba2hi commented on sdk/src/main/java/velocitas/Utils.kt:
```diff
@@ -0,0 +1,71 @@
+package velocitas
```
package name to be discussed. 

I could not find anything comparably in the cpp one, only the namespace "velocitas": https://github.com/eclipse-velocitas/vehicle-app-cpp-sdk/blob/main/sdk/include/sdk/grpc/AsyncGrpcFacade.h#L24

wba2hi replied:
decision: org.eclipse.velocitas.sdk

wba2hi commented on sdk/build.gradle.kts:
```diff
@@ -0,0 +1,54 @@
+import com.google.protobuf.gradle.id
+
+plugins {
+    id("java-library")
```
sdk is setup as a java library as per discussions - not an Android Library!

wba2hi commented on sdk/src/main/java/velocitas/sdk/VehicleApp.kt:
```diff
@@ -0,0 +1,47 @@
+
+        onStart()
+
+        // TODO: Fix busy waiting
+        while (true) {
+            Thread.sleep(1)
+        }
```
Took it over from here: https://github.com/eclipse-velocitas/vehicle-app-cpp-sdk/blob/main/sdk/include/sdk/VehicleApp.h
However I don't know what purpose it aims for

wba2hi commented on sdk/src/main/java/velocitas/sdk/grpc/GrpcClient.kt:
```diff
@@ -0,0 +1,44 @@
+            }
+        }
+
+        // static auto isComplete = [](const auto& activeCall) { return activeCall->m_isComplete; };
+
+        // {
+        //    std::scoped_lock<std::mutex> lock(m_mutex);
+        //    m_activeCalls.erase(std::remove_if(m_activeCalls.begin(), m_activeCalls.end(), isComplete),
+        //        m_activeCalls.end());
+        // }
```
Didn't totally understand the syntax here, but I think I covered the methods behavior: https://github.com/eclipse-velocitas/vehicle-app-cpp-sdk/blob/main/sdk/src/sdk/grpc/GrpcClient.cpp#L30

Forgot to remove the lines

Chrylo commented on .idea/vcs.xml:
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="VcsDirectoryMappings">
+    <mapping directory="$PROJECT_DIR$/.." vcs="Git" />
```
Got an error because of this line. Please check which .idea files to push into git.

wba2hi replied:
Could you shortly show the error? 

I basically used the default Android Studio .gitignored merged together with the one from kuksa-sdk which is based on an .gitignore generator

Here you can also find the recommendations from idea: 
https://intellij-support.jetbrains.com/hc/en-us/articles/206544839-How-to-manage-projects-under-Version-Control-Systems => https://github.com/github/gitignore/blob/main/Global/JetBrains.gitignore 
and these also do not exclude .idea/vcs.xml

wba2hi replied:
Problem solved

wba2hi committed:
Add Apache-2.0 License and SPDX Identificator to Files

Chrylo commented on app/build.gradle.kts:
AGP 8.5.1 is release instead of using the RC version. :)

wba2hi committed:
Remove Unused Imports

wba2hi committed:
Extract Dependencies to .toml and Update Versions

wba2hi committed:
Add KtLint

wba2hi committed:
Fix KtLint Issues

Chrylo commented on sdk/src/main/java/velocitas/Utils.kt:
```diff
@@ -0,0 +1,71 @@
+import java.net.URI
+import java.net.URL
+
+fun getEnvVar(varName: String, defaultValue: String = ""): String {
```
Can we use a system extension?

wba2hi replied:
Moved to Middleware

Chrylo commented on sdk/src/main/java/velocitas/Utils.kt:
```diff
@@ -0,0 +1,71 @@
+ * ("login") part of the URL.
+ * Other elements to be added as needed ...
+ */
+class SimpleUrlParse(url: String) {
```
parse -> parseR?

Chrylo commented on sdk/src/main/java/velocitas/Utils.kt:
Let us not start with utils :)

We can extract the parser to some parser package and rename the class to the parser one.

Chrylo commented on sdk/src/main/java/velocitas/Utils.kt:
```diff
@@ -0,0 +1,71 @@
+        private set
+
+    init {
+        parse(url)
```
implicit logic in constructor. Let the caller use the parse method. Return a data class maybe. Remove the url from the constructor.

Chrylo commented:
Required:
- Readme disclaimer states that this repo is “work in progress / incubation status”

Chrylo commented on sdk/src/main/java/velocitas/sdk/AsyncResult.kt:
```diff
@@ -0,0 +1,118 @@
+    /**
+     * Inserts the result and notifies any waiters.
+     *
+     * @param result  Result to insert.
```
Best practice from google was to not use @param anymore but just reference it in sentences. ;)

wba2hi replied:
Should be fixed by now

Chrylo commented on sdk/src/main/java/velocitas/sdk/AsyncResult.kt:
```diff
@@ -0,0 +1,118 @@
+     */
+    fun await(): TResultType {
+        if (resultCallback != null) {
+            throw RuntimeException("Invalid usage: Either call await() or register an onResult callback!")
```
I don't understand the message here: "Call await()"? but the dev did call await?

wba2hi replied:
It's a generic message, which says that the dev EITHER can call await() to block until a result is received OR register an async resultCallback. In this case here the resultCallback is null, so he already set an async listener, therefore calling await afterwards is inconsistent. 

Same message is also thrown when he calls first await and then sets a resultCallback

Chrylo commented on sdk/src/main/java/velocitas/sdk/AsyncResult.kt:
```diff
@@ -0,0 +1,118 @@
+class VoidResult
+
+class AsyncResult<TResultType> {
+    private var result: TResultType? = null
```
Is it necessary that the result is optional after the construction? Could the bis a constructor param?

wba2hi replied:
Yes, it needs to be optional, you can't influence the result from the Server. Server obviously could just send out an error message instead of an actual result 

AsyncResult - even though the name suggest so - is not a simple Data class, but it does more, since it also provides methods like await() and onResult() - so just simply moving it to the constructor won't work. However, I agree that the whole class is a really weird construct!

Chrylo replied:
Accepted for now, the whole constructs needs a refactoring at some point.

Chrylo commented on sdk/src/main/java/velocitas/sdk/AsyncResult.kt:
```diff
@@ -0,0 +1,118 @@
+     */
+    fun insertResult(result: TResultType) {
+        try {
+            if (resultCallback != null) {
```
Do we need this if check? "?" should already handle it.  Then only the result assignment needs an "if" if even necessary. Same in other methods.

wba2hi replied:
Again some weird design decision but basically this method here EITHER calls the resultCallback if it is set, or set's the AsyncResult#result if the user is await()-ing for the result. If the resultCallback is set the #result field is not set. 

This is to be consistent with the previous behavior, that we throw an Exception when the user sets an async listener AND tries to await the result.

Chrylo commented on sdk/src/main/java/velocitas/sdk/AsyncResult.kt:
```diff
@@ -0,0 +1,118 @@
+
+    /**
+     * Calls the specified callback when the result is available.
+     *        The callback invocation is done by a worker thread.
```
Weird space here

Chrylo commented on sdk/src/main/java/velocitas/sdk/Job.kt:
```diff
@@ -0,0 +1,59 @@
+
+import java.util.concurrent.atomic.AtomicBoolean
+
+interface IJob {
```
We need better naming than "I" prefixes ;)

Maybe "ExecuterJob" and below "LambdaJob" + "ReoccuringLambdaJob"?

Chrylo commented on sdk/src/main/java/velocitas/sdk/Logger.kt:
```diff
@@ -0,0 +1,115 @@
+}
+
+object Logger {
+    private var impl: ILogger = ConsoleLogger()
```
A logger is not a logger but contains another interface logger? I am confused.^^ Do we even need this one?

wba2hi replied:
Well, let's call it an uncommon strategy pattern. I don't think the abstraction is bad per-say. You can now easily also implement a "FileLogger" or other stuff, because it's quite flexible. 

Do we need it? Depends on the requirements! 
Do we need it now? Maybe not - but as stated before it was hard to make a cut on what to take over into the sdk and what not.

Chrylo commented on sdk/src/main/java/velocitas/sdk/Logger.kt:
```diff
@@ -0,0 +1,115 @@
+
+import java.util.Date
+
+interface ILogger {
```
"VelocitasLogger" or just "Logger"?

Chrylo commented on sdk/src/main/java/velocitas/sdk/Status.kt:
```diff
@@ -0,0 +1,36 @@
+ *
+ * @constructor Construct a new Status object without any errors.
+ */
+class Status() {
```
Should this be a data class?

Chrylo commented on sdk/src/main/java/velocitas/sdk/ThreadPool.kt:
```diff
@@ -0,0 +1,53 @@
+        }
+    }
+
+    fun enqueue(job: IJob) {
```
No doc

Chrylo commented on sdk/src/main/java/velocitas/sdk/ThreadPool.kt:
```diff
@@ -0,0 +1,53 @@
+    constructor() : this(2)
+
+    private val workerThreads = Vector<Thread>()
+    private val jobs = ArrayBlockingQueue<IJob>(1000)
```
Magic number

Chrylo commented:
Implementing Detekt would be great. But we can do that in a follow up task. But we should not wait too long for that.

Chrylo commented on sdk/src/main/java/velocitas/sdk/ThreadPool.kt:
```diff
@@ -0,0 +1,53 @@
+    private fun threadLoop() {
+        while (isRunning.get()) {
+            val polledJob = jobs.poll()
+            if (polledJob != null) {
```
Nested complexity, early return?

wba2hi replied:
early-return is not good in this case as it would jump break out of the while loop and therefore stop the threads therefore an early-continue is required here.

Chrylo commented on sdk/src/main/java/velocitas/sdk/VehicleApp.kt:
```diff
@@ -0,0 +1,47 @@
+ */
+abstract class VehicleApp
+{
+
```
Empty line

Chrylo commented on sdk/src/main/java/velocitas/sdk/VehicleApp.kt:
```diff
@@ -0,0 +1,47 @@
+     */
+    fun run() {
+        Logger.info("Running App...")
+        Middleware.getInstance().start()
```
Probably from the c++ variant taken, but can we do it better? 

Does it have to be a singleton? If yes I am wondering why we are not using the SingletonHolder pattern that was introduced here. :)

wba2hi replied:
> Probably from the c++ variant taken ...

Yes: https://github.com/eclipse-velocitas/vehicle-app-cpp-sdk/blob/main/sdk/src/sdk/VehicleApp.cpp#L41

> ... but can we do it better?

Since this would be a breaking change, this is to be discussed with the team

> ... If yes I am wondering why we are not using the SingletonHolder pattern that was introduced here. :)

That's a valid point :D

Chrylo replied:
Will be discussed

wba2hi replied:
Implemented a MiddlewareFactory which will instantiate the Middleware

Chrylo commented on sdk/src/main/java/velocitas/sdk/grpc/GrpcCall.kt:
```diff
@@ -0,0 +1,76 @@
+ * @param TRequestType The data type of the request.
+ * @param TReplyType   The data type of the reply.
+ */
+class GrpcSingleResponseCall<TRequestType, TReplyType> : GrpcCall() {
```
SingleResponseGrpcCall right? 

A "HawaiiToast" would also not be splitten up into "HawaiiSpecialVeggiToast" but "SpecialVeggiHawaiiToast" ;)

Same with the other naming.

Chrylo commented on sdk/src/main/java/velocitas/sdk/grpc/GrpcCall.kt:
```diff
@@ -0,0 +1,76 @@
+    var request: TRequestType? = null
+        private set
+
+    private val readReactor: ReadReactor<TRequestType, TReplyType> = ReadReactor(this)
```
Circular references?

wba2hi replied:
Fixed. A bigger refactoring was done here - I think it should be fine, since it mostly contains internal / grpc specific stuff. 
I don't exactly know why we need the GrpcCall, because basically the only thing is does, is tracking the active requests / responses and the complete state for which is tracked inside the GrpcClient. But let's keep it for now (maybe something to be discussed inside the Architecture round, what's the purpose of this thingy)

It's still pretty rough around the edges, but all in all I'm a satisfied how the GrpcCall / AsyncBrokerGrpcFacade now handles the logic.

Chrylo commented on sdk/src/main/java/velocitas/sdk/middleware/Middleware.kt:
```diff
@@ -0,0 +1,91 @@
+    /**
+     * Triggers the start of the middleware
+     */
+    open fun start() {
```
```suggestion
    open fun start() { }
```

Or should this method be abstract?

wba2hi replied:
If I understood this correctly, then the CPP SDK provides empty methods as default, correct me if I'm wrong: 
https://github.com/eclipse-velocitas/vehicle-app-cpp-sdk/blob/main/sdk/include/sdk/middleware/Middleware.h#L59

wba2hi committed:
Implement Detekt

wba2hi committed:
Fix Detekt Findings

wba2hi committed:
Add Incubation Status Note

wba2hi committed:
Extract SimpleUrlParser into separate Class

wba2hi committed:
Refactor Logger Namings

wba2hi committed:
Use SingletonHolder for Middleware Singleton

wba2hi committed:
Make Status a data class

wba2hi committed:
Rename Interface IJob to ExecutorJob

wba2hi committed:
Remove Util Class

wba2hi committed:
Improve KDoc

wba2hi committed:
Remove getEnvVar Method

wba2hi committed:
Add Github Action Workflow for PRs

wba2hi commented:
Added a PR Pipeline but it looks like it is now "to late" for it to be recognized. 
We might have to open a new PR for it to be recognized?

wba2hi committed:
Use kuksa.val.v2 Protocol

wba2hi committed:
Implementation of the AsyncBrokerGrpcFacade

wba2hi committed:
Make Subscriptions cancellable

Chrylo commented on sdk/src/main/java/velocitas/sdk/middleware/Middleware.kt:
```diff
@@ -0,0 +1,87 @@
+        val envVar = System.getenv(TYPE_DEFINING_ENV_VAR_NAME) ?: ""
+        val middlewareType = envVar.lowercase()
+        if (middlewareType.isEmpty()) {
+            NativeMiddleware()
```
Do we need all if checks?  NativeMiddleware() is the same for 2 checks

wba2hi replied:
Replaced with a when and merged both branches

Chrylo commented on sdk/src/main/java/velocitas/sdk/grpc/GrpcClient.kt:
```diff
@@ -0,0 +1,58 @@
+    private val recurringJob: RecurringJob
+    private val activeCalls = mutableListOf<GrpcCall>()
+
+    val numActiveCalls: Int
```
activeCallsCount?

Chrylo commented on sdk/src/main/java/velocitas/sdk/middleware/NativeMiddleware.kt:
```diff
@@ -0,0 +1,77 @@
+ */
+class NativeMiddleware : Middleware(TYPE_ID) {
+
+    override fun getServiceLocation(serviceName: String): String {
```
Ok, this is a crazy get.^^ Maybe it should really be a "find" or similar.

Chrylo reviewed (changes requested):
I don't get the whole picture here so I mainly reviewed it on code level and not functional. An architecture design discussion is ongoing for some classes here. However it would be good if someone additional can approve if this goes into the right direction.

wba2hi committed:
Minor Renamings

wba2hi committed:
Replace If-Else Cascade with When

wba2hi committed:
Fix Lint Issues

wba2hi committed:
Remove Singletons

wba2hi committed:
Refactor Logger

- Rename VelocitasLogger back to Logger.
- Rename interface Logger to LoggingStrategy
- Extract LoggingStrategy and ConsoleLoggingStrategy to separate classes

wba2hi committed:
Change Package to "org.eclipse.velocitas.sdk"

wba2hi committed:
Rename SimpleUrlParser to UrlParser

wba2hi committed:
Refactor ThreadPool to use ScheduledExecutorService

wba2hi committed:
Use AsyncResult for AsyncBrokerGrpcFacade

wba2hi commented on sdk/src/main/java/org/eclipse/velocitas/sdk/grpc/AsyncBrokerGrpcFacade.kt:
```diff
@@ -0,0 +1,438 @@
+     *    NOT_FOUND if the requested signal doesn't exist
+     *    PERMISSION_DENIED if access is denied
+     */
+    fun getValue(signalId: SignalID): AsyncResult<GetValueResponse> {
```
I changed these calls to AsyncResult for a better user experience. User can now decide if he wants to call the corresponding gRPC method in a synchronous or asynchronous way. However, right now this has one drawback:
- the visibility of AsyncResult#insertResult and AsyncResult#insertError is public right now; so could be abused by a developer

Probably needs to be changed to internal but this will generate these ugly methods on java side (and will basically still be accessable from it). Want to hear your opinion on it.

I really hate that kotlin still does not plan to support package privacy

wba2hi committed:
Add ThreadPool Cancelation Logic

wba2hi committed:
Add '.idea/vcs.xml' to .gitignore

doosuu commented on app/.gitignore:
missing trailing newline

doosuu commented on app/proguard-rules.pro:
```diff
@@ -0,0 +1,21 @@
+
+# If you keep the line number information, uncomment this to
+# hide the original source file name.
+#-renamesourcefileattribute SourceFile
```
Missing trailing newline

doosuu commented on app/.gitignore:
entire app directory (which is an android app) really needed for a pure kotline based template?

wba2hi replied:
I'm unsure if I understand the question here correct, I assume you don't want to use the subfolder "app" for the application but just the project root?
=> In theory that might be possible, best practice on Android side is to use the subfolder so it's easy to extend in case you need another submodule or separate app module. 

However if your question is, if we need to check in the app directory into source control in this ticket?
=> then no we could also remove it and add it once we implement it inside the template ticket.

doosuu replied:
My point is that this is a plain Kotlin template, not an android app. The app directory has android specific content.

wba2hi replied:
Understood - you are right! We probably need both, one "plain kotlin app" and one "android automotive app" to test stuff. I removed the app folder now and will re-introduce it correctly configured the follow-up ticket

doosuu commented on sdk/src/main/java/org/eclipse/velocitas/sdk/concurrency/AsyncResult.kt:
We spoke about this - can we remove the `AsyncResult` completely and use a `Future` instead?

wba2hi replied:
Replaced them with ListenableFutures, removed a lot of complexity which is a good thing, however somehow it feels like GrpcClient (now a marker interface) and the whole BrokerGrpcFacade are now simple wrappers

doosuu commented on sdk/src/main/java/org/eclipse/velocitas/sdk/parser/UrlParser.kt:
```diff
@@ -0,0 +1,80 @@
+ * ("login") part of the URL.
+ * Other elements to be added as needed ...
+ */
+class UrlParser {
```
Is there no library in kotlin to do this?

wba2hi replied:
Well,... depends on the requirements. We can probably use [URI](https://docs.oracle.com/javase/8/docs/api/java/net/URI.htm) or even better [URL](https://docs.oracle.com/javase/8/docs/api/java/net/URL.html) classes, however these do not work with classes which don't have a leading scheme, [a feature which the UrlParser actively advertises with
](https://github.com/eclipse-velocitas/vehicle-app-cpp-sdk/blob/main/sdk/include/sdk/Utils.h#L88)

So, if we are good with always specifying URLs with `http://localhost:12345` instead of `localhost:12345` we can go for it, otherwise we will need to use the custom implementation.

doosuu replied:
Let's stick to the current implementation to be in line with the other languages.

doosuu commented on sdk/src/main/java/org/eclipse/velocitas/sdk/concurrency/Job.kt:
Jobs and thread pool should not be needed since we have [coroutines](https://kotlinlang.org/docs/coroutines-overview.html#how-to-start).

Both classes just exist in c++ to mimic coroutines without actual language support (or using boost) in C++ < 20

wba2hi replied:
Removed

doosuu commented on sdk/src/main/java/org/eclipse/velocitas/sdk/grpc/GrpcCall.kt:
```diff
@@ -0,0 +1,96 @@
+ * @param TRequestType The data type of the request.
+ * @param TResponseType The data type of the response.
+ */
+class SingleResponseGrpcCall<TRequestType, TResponseType>(
```
These are just here b/c of gRPC's lacking async API in C++ - consider using something like https://github.com/rouzwawi/grpc-kotlin instead. Let's try to remove these classes.

wba2hi replied:
Added official implementation for https://github.com/grpc/grpc-kotlin. Henceforth the API of BrokerGrpcFacade also changed from ListenableFutures to suspend functions and Flows.

doosuu commented:
Consider adding pre-commit with similar settings to the other repos

wba2hi committed:
Add missing newLines to EOF

wba2hi committed:
Remove Jobs and ThreadPool Logic

wba2hi committed:
Refactor BrokerGrpcFacade to return ListenableFutures

wba2hi committed:
Use gRPC Kotlin

Chrylo commented:
> Consider adding pre-commit with similar settings to the other repos

We could use "husky" to enforce the commit hooks. I would probably do that in another ticket.

Chrylo commented:
> I don't get the whole picture here so I mainly reviewed it on code level and not functional. An architecture design discussion is ongoing for some classes here. However it would be good if someone additional can approve if this goes into the right direction.

LGTM to make a cut here for the initial implementation

wba2hi committed:
Remove App

doosuu reviewed (approved):
lgtm
